---
title: "Data analyses workflow"
subtitle: "Part 1: quality fltering and rarefaction"
output:
  word_document:
    toc: yes
  html_notebook:
    toc: yes
  html_document:
    highlight: tango
    number_sections: yes
    theme: readable
    toc: yes
editor_options:
  chunk_output_type: console
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r load-packages, include=FALSE}
pacman::p_load(rmarkdown,knitr,phyloseq, ggplot2, vegan, gridExtra, vegan, plyr, dplyr, microbiome, RColorBrewer, tidyr, ggcorrplot, readxl)

```

# Read OTU table 

```{r read_table, include=FALSE}
setwd("C:/Users/ccasas/OneDrive - Facultad de Agronomía - Universidad de Buenos Aires/Activos/2021-nov OTU Table and results")

otu <- read_excel("Cecilia_OTU_table_updatedBLAST_copy.xlsx", sheet = "OTU table")

# Change names for convenience
colnames(otu)[2] <- "AccNoUNITEv9_6"
colnames(otu)[14:25]<-c("e_value" , "query_len", "query_start", "query_end", "target_len", "target_start", "target_end",  "align_len", "identities", "gaps", "coverage_perc", "id_perc" )

# names(otu)
```

# Keep only soil samples

The table contains:

-   Soil samples (CC001 - CC238) \* Seeds samples (1_Seed_No_St 1_Seed_St)

-   ITS4ngsUni001 ITS4ngsUni003 ITS4ngsUni008 ITS4ngsUni014 ITS4ngsUni015 1_ITS4ngsUni_089 2_ITS4ngsUni_087 2_ITS4ngsUni_088 3_ITS4ngsUni_087 3_ITS4ngsUni_088 3_ITS4ngsUni_089 4_ITS4ngsUni_087 4_ITS4ngsUni_088

-   NC1 NC2 NC3 NC4

-   Pos1 Pos2 Pos3 Pos4

-   empty1 empty2

I exclude all columns except from Soil Samples

```{r, include= FALSE}
exclude<-c("ITS4ngsUni001",	"ITS4ngsUni003",	"ITS4ngsUni008",	"ITS4ngsUni014",	"ITS4ngsUni015",	"1_ITS4ngsUni_089",	"1_Seed_No_St",	"1_Seed_St",	"2_ITS4ngsUni_087",	"2_ITS4ngsUni_088",	"3_ITS4ngsUni_087",	"3_ITS4ngsUni_088",	"3_ITS4ngsUni_089",	"4_ITS4ngsUni_087",	"4_ITS4ngsUni_088",	"NC1",	"NC2",	"NC3",	"NC4",	"Pos1",	"Pos2",	"Pos3",	"Pos4",	"empty1",	"empty2")

otu_soil<-otu[,!names(otu)%in%exclude]

```

# Read Phyloseq

## OTU table
```{r}
colsCC<-grep("CC", names(otu_soil), value = TRUE)

otu_file<-otu_soil[,names(otu_soil)%in%colsCC]
rownames(otu_file)<-otu_soil$OTU_ID...27
dim(otu_file)

```

## Taxonomy file
```{r}
taxonomy_file<-otu_soil[, c("Kingdom","Phylum","Class","Order",	"Family","Genus","Species")]
rownames(taxonomy_file)<-otu_soil$OTU_ID...27

```

## Sample data
```{r}
site_data<-read.csv("Data FE-2021-00537.csv", header= TRUE, sep= ";", row.names= colsCC)
#rownames(site_data)
site_data$Herbivory<-as.factor(site_data$Herbivory)
#summary(site_data)

```


## Read the files as phyloseq class object.

```{r}
otu_file<-as.matrix(otu_file)
taxonomy_file<-as.matrix(taxonomy_file)

OTU = otu_table(otu_file, taxa_are_rows = TRUE)
TAX = tax_table(taxonomy_file)
site_data = sample_data(site_data)

phylo.data = phyloseq(OTU, TAX, site_data)
phylo.data

```

## Remove zero OTUS 

Zero reads correspond to OTU detected in samples excluded above such as Seeds, ITS4, NC, pos, empty, etc...

```{r}
phylo.data<-subset_taxa(phylo.data, taxa_sums(phylo.data)>0)  
phylo.data

```

### Save phylo.data

```{r}
saveRDS(phylo.data, file = "2024-01-13 phylo.data.rds")

```


# Quality filtering

## Rarefaction: define threshold 
```{r}
n_seq <- colSums(otu_table(phylo.data))
# Create a data frame with sample names and corresponding sequencing depths
depth_data <- data.frame(Sample = colnames(otu_table(phylo.data)), n_seq = n_seq)

depth_data %>%
  ggplot(aes(x= 1, y=n_seq)) +
  geom_jitter() +
  scale_y_log10()

depth_data %>%
  arrange(n_seq) %>%
  ggplot(aes(x= 1:nrow(.), y=n_seq)) +
  geom_line() 

depth_data %>%
  arrange(n_seq) %>%
  ggplot(aes(x= 1:nrow(.), y=n_seq)) +
  geom_line() +
  geom_point()+
  coord_cartesian(xlim=c(0,50), ylim=c(0,25000)) +
  geom_hline(yintercept=10000, color="grey", linetype='dashed', size=1)


depth_data_arran<-depth_data %>%
  arrange(n_seq) 

head(depth_data_arran, n = 20)

```


 Sample SequencingDepth
  CC122               5
  CC187            2746
  CC123            5201
  CC189            5270
  CC188            7526
  CC024            8600 # DEPTH for rarefaction = 10000
  CC140           10397
  CC126           10790
  CC118           11830
  CC231           12691
  CC119           13351
  CC115           13408
  CC069           14542
  CC190           14676
  CC163           14790
  CC166           14935
  CC023           15116
  CC208           15270
  CC233           15336
  CC226           15877

> The same with seqeuncing depth calculated from demultiplexed R1 fastq files. 

This section is only internal to check that we arrive to the same conclusion about the rarefaction threshold. 

```{r}
# Create a data frame with sample names and corresponding sequencing depths
depth_data <- data.frame(Sample = rownames(sample_data(phylo.data)), seq_d = sample_data(phylo.data)$R1_seq_depth)

# depth_data %>%
#   arrange(seq_d)

depth_data %>%
  ggplot(aes(x= 1, y=seq_d)) +
  geom_jitter() +
  scale_y_log10()

depth_data %>%
  arrange(seq_d) %>%
  ggplot(aes(x= 1:nrow(.), y=seq_d)) +
  geom_line() 

depth_data %>%
  arrange(seq_d) %>%
  ggplot(aes(x= 1:nrow(.), y=seq_d)) +
  geom_line() +
  geom_point() +
  coord_cartesian(xlim=c(0,50), ylim=c(0,10000000)) +
  geom_hline(yintercept=3000000, color="grey", linetype='dashed', size=1)


depth_data_arran<-depth_data %>%
  arrange(seq_d) 

head(depth_data_arran, n = 20)

```

   Sample   seq_d
1   CC122   43487
2   CC187 1041912
3   CC189 1914596
4   CC123 2024816
5   CC188 2738168
6   CC024 3338177    DEPTH for rarefaction = 3000000
7   CC126 3610335
8   CC140 3765785
9   CC118 4088419
10  CC231 4613673
11  CC115 4672192
12  CC119 4894737
13  CC190 5096875
14  CC069 5125647
15  CC233 5193437
16  CC023 5385127
17  CC163 5442813
18  CC166 5482941
19  CC226 5534694
20  CC208 5647867

•	**Question**: Should I include or exclude CC024?

# Sample coverage 
## Calculate good coverage at sample level
### Sample singletons
```{r}
n_seq <- colSums(otu_table(phylo.data))
seq_single <- colSums(otu_table(phylo.data)==1)

depth_data <- data.frame(Sample = colnames(otu_table(phylo.data)), 
                         n_seq = n_seq, 
                         seq_single = seq_single)

coverage_stat<-depth_data %>%
  reframe(n_seq = n_seq, 
          goods= 100*(1 - seq_single/n_seq)) 

coverage_stat%>%
  ggplot(aes(x=n_seq, y=goods))+
  geom_point()

# coverage_stat %>%
#   arrange(goods)

```

> coverage is 100% for all sample....there are no singletons at sample level. 

### Sample doubletons coverage
```{r}
# Calculate good coverage

depth_data$seq_double <- colSums(otu_table(phylo.data)==2)

coverage_stat<-depth_data %>%
  reframe(Sample= rownames(depth_data), 
          n_seq = n_seq, 
          goods= 100*(1 - seq_double/n_seq)) 

coverage_stat%>%
  ggplot(aes(x=n_seq, y=goods))+
  geom_point()

coverage_stat %>%
  filter(n_seq>10000) %>% 
  ggplot(aes(x=n_seq, y=goods))+
  geom_point()


coverage_stat %>%
  filter(goods>97) %>% 
  ggplot(aes(x=n_seq, y=goods))+
  geom_point()

head(coverage_stat %>%
  arrange(goods))

```

   Sample            n_seq    goods
1    CC122               5 80.00000
2    CC187            2746 96.75892
3    CC189            5270 97.79886
4    CC123            5201 98.13497
5    CC188            7526 98.25937
6    CC140           10397 98.45148

> Except from the first (and maybe the second) coverage of doubletons is also good. 
> Then, **may threshold be 5000 instead of 10000?**

> The difference between both thresholds is 3 samples.
> I will keep to 10000 so I have more depth in the analyses and I am not missing many samples (6)

One thing to **observe** is that from the six samples, 5 came from two sites
But I have 8 plants per site. 

```{r}
sample_data(phylo.data)[rownames(sample_data(phylo.data)) %in% c("CC122", "CC187", "CC123", "CC189", "CC188", "CC024"), c("Sample", "Site", "PrecWorldClim")]

```

      Sample Site    MAP  Environment
CC024  CC024    3    307  semiarid
CC122  CC122   18    160  arid
CC123  CC123   18    160  arid
CC187  CC187   26    442  semiarid
CC188  CC188   26    442  semiarid
CC189  CC189   26    442  semiarid

MAP = mean annual precipitation (mm)

```{r}
range(site_data$PrecWorldClim)  
hist(site_data$PrecWorldClim)  
abline(v=160, col="red")
abline(v=307, col="blue")
abline(v=442, col="blue")

```

# Rarefaction curve

```{r}
OTU<-otu_table(phylo.data)
OTU_tras<- as.data.frame(t(OTU))

rowSums(t(OTU))[order(rowSums(t(OTU)), decreasing = T)]
S <- specnumber(t(OTU))
raremax <- min(rowSums(t(OTU)))

```

```{r}
## plotting the rarefaction curve, by sample
# pdf("Rarefaction curve by sample.pdf")
# rarecurve(OTU_tras, step = 20, label = F, col = "gray", legend=T)
# dev.off()

#plotting rarecurve  by grazing
#create color vector
# rare_landuse <- rare_color <- site_data$Herbivory[match(row.names(OTU_tras), row.names(site_data))]

# rare_color<-revalue(rare_color, c("Intense"="red", "Moderate"="blue"))

#pdf("Rarefaction curve by grazing.pdf")
#rarecurve(forrare, step = 20, label = F, col = rare_color, legend=T)
#dev.off()
```


## Octave plot

```{r}
OTUsum<-rowSums(t(OTU)) 
sum(OTUsum)
length(OTUsum)

mod.oct <- prestonfit(OTUsum, tiesplit=F)
plot(mod.oct)

```

## Rarefaction

>I am using sample.size= 10000 according to previous figures

```{r}
threshold= 10000

phylo.rare <- rarefy_even_depth(phylo.data, sample.size = threshold, rngseed = 1111 , replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

# ...
# 6 samples removed because they contained fewer reads than `sample.size`.
# Up to first five removed samples are: 
# 
# CC024
# CC122
# CC123
# CC187
# CC188
# ...

thrown_samples <- names(colSums(OTU)[colSums(OTU)<threshold])

write.csv(
  data.frame(sample.name = thrown_samples,
             site = site_data$Site[match(thrown_samples, row.names(site_data))],
             plant = site_data$Plant[match(thrown_samples, row.names(site_data))],
             sample.depth = unname(colSums(OTU)[colSums(OTU)<threshold])),
  "2024-01-12 thrown samples after rarefaction.csv")

```

# Save phyloseq object of rarefied OTU
```{r}
saveRDS(phylo.rare, file = "2024-01-13 phylo.data.after.rarefaction.rds")
```

# Rarefaction formula by Schloss and comparison

```{r}
rare_schloss_fn <- function(x, sample){
  
  x <- x[x>0]
  sum(1-exp(lchoose(sum(x) - x, sample) - lchoose(sum(x), sample)))
  
}
```

```{r}
# richness from original data set
depth_data$rich_obs<- estimate_richness(phylo.data, measures = "Observed")[,1]

# richness from rarefied data to compare with schloss formula
rich_rare<- estimate_richness(phylo.rare, measures = "Observed")
rich_rare$Sample=rownames(rich_rare)
colnames(rich_rare)[1]<-"rich_rare"
head(rich_rare)

depth_data<-merge(depth_data, rich_rare, by="Sample", all=T)
#head(depth_data)

depth_data %>%
  summarize(min=min(n_seq), 
            max=max(n_seq))
120358/5


```

## prerate data set to apply rarefaction by Schloss
```{r}
OTU<-as.data.frame(otu_table(phylo.data))
OTU$otu_name<-rownames(OTU)

otu_piOTUotu_pivot <- pivot_longer(OTU, cols=starts_with("CC"), names_to="Sample", values_to="n_seq", cols_vary="fastest") %>%
  as.data.frame()

head(otu_piOTUotu_pivot)

```

## rarefaction by Schloss
```{r}
rarefaction_schloss<-otu_piOTUotu_pivot %>%
    dplyr:: group_by(Sample) %>%
    dplyr::summarise(rich_schloss= rare_schloss_fn(n_seq, threshold), 
              n=sum(n_seq), 
              sobs= sum(n_seq>0))

depth_data<-merge(depth_data, rarefaction_schloss, by="Sample", all=T)
head(depth_data)

```

#### Plots
```{r}
depth_data %>%
  ggplot(aes(x=rich_rare, y=rich_schloss)) +
  geom_point() +
  geom_smooth()


depth_data %>%
 pivot_longer(cols=c(rich_obs, rich_rare, rich_schloss), values_to = "rich", names_to = "type") %>%
  ggplot(aes(x=n, y=rich, col=type))+
  geom_point()+
  geom_smooth()

```

# Filtering after rarefaction

## Chimeras: Singletons and low number of reads

*  We removed singletons.       
*  We removed doubletons with similarity <0.9 and similarity >0.9&coverage <0.98. 
*  and OTUs with BLASTn e-value >10-50 to exclude additional potential chimeras. 

In this OTU table, *Similarity = id_perc (id%)*

### Remove singletons 

```{r}
#subset_taxa(phylo.rare, taxa_sums(phylo.rare) == 0)

subset_taxa(phylo.rare, taxa_sums(phylo.rare) == 1)

phylo.rare.clean<-subset_taxa(phylo.rare, taxa_sums(phylo.rare) > 1)  
#phylo.rare.clean

# taxa_sums(phylo.rare.clean)
```

There were 5922 singletons.

##### Exclude doubletons. 
We excluded doubletons with similarity <0.9 and similarity >0.9&coverage <0.98. 

```{r, include=FALSE}

double_names<-taxa_names(subset_taxa(phylo.rare.clean, taxa_sums(phylo.rare.clean) ==2))

length(double_names)

to.exclude.id<-otu[otu$OTU_ID...1 %in% double_names,] %>%
   select (OTU_ID...1, id_perc,  coverage_perc, query_len) %>%
   filter(id_perc<.9)

#plot(to.exclude.id$id_perc)

to.exclude.cov<-otu[otu$OTU_ID...1 %in% double_names,] %>%
  select (OTU_ID...1, id_perc,  coverage_perc, query_len) %>%
   filter(id_perc>.9) %>%
   filter(coverage_perc<.98)

#plot(to.exclude.cov$coverage_perc)
  
to.exclude<-merge(to.exclude.id, to.exclude.cov, all=TRUE)

dim(to.exclude.id)
dim(to.exclude.cov)
dim(to.exclude)

```

> There are `r  dim(to.exclude.id)[1]+dim(to.exclude.cov)[1]` reads of OTUs in soil samples after excluding OTUs with == 2 reads (doubletons) with  id_perc\<=.9 or id_perc\>=.9 + coverage_perc\<=.98 in total (row sum).\

```{r}
filtered_otu <- subset(otu_table(phylo.rare.clean), !rownames(otu_table(phylo.rare.clean)) %in% to.exclude$OTU_ID...1)
phylo.filtered <- merge_phyloseq(filtered_otu, tax_table(phylo.rare.clean), sample_data(phylo.rare.clean))
phylo.filtered

```

# Save phylo.data
```{r}
saveRDS(phylo.filtered, file = "2024-01-13 phylo.data.rarefaction.filtered.rds")
```


```{r}
OTUsum<-as.integer(taxa_sums(phylo.filtered))
str(OTUsum)
mod.oct <- prestonfit(OTUsum, tiesplit = F)
plot(mod.oct)

sum(OTUsum) 
length(OTUsum) 

```

<!-- ## see how many have less than 5 occurences -->

```{r, include= FALSE}
# phylo.rare.clean.less.5<-subset_taxa(phylo.filtered, taxa_sums(phylo.filtered) < 5)  
# phylo.rare.clean.less.5

```

# Select only kingdom Fungi

```{r, include=FALSE}
phylo.fungi <- subset_taxa(phylo.filtered, Kingdom=="Fungi")
phylo.fungi

```

<!-- ## see how many have less than 5 occurences -->

<!-- ```{r, include= FALSE} -->
<!-- phylo.fungi.less.5<-subset_taxa(phylo.fungi, taxa_sums(phylo.rare) < 5)   -->
<!-- phylo.fungi.less.5 -->

<!-- ``` -->

## Save phylo.fungi
```{r}
saveRDS(phylo.fungi, file = "2024-01-13 phylo.data.after.rarefaction.fungi.rds")
```

## Octave plot

```{r}

OTUsum<-as.integer(taxa_sums(phylo.fungi))
str(OTUsum)
mod.oct <- prestonfit(OTUsum, tiesplit = F)
plot(mod.oct)

sum(OTUsum) 
length(OTUsum) 
```


#### just to check = proportion of OTUs with 100% similarity
```{r}

filtered.otu<-rownames(otu_table(phylo.fungi))

filtered.qualy<-otu[otu$OTU_ID...1 %in% filtered.otu,] %>%
  select (OTU_ID...1, id_perc,  coverage_perc, query_len, e_value) 
   
hist(filtered.qualy$id_perc)

length(which(filtered.qualy$id_perc>=.90))/nrow(filtered.qualy)

```

## e-value

We considered e-values of BLASTn >= 1e−50 reliable to assign sequences to the fungal kingdom.
(E-values between 1e−20 and 1e−50 were manually checked against the 10 bestmatches for accurate assignment.
whereas those e-values <1e−20 were considered "unknown.")

```{r}
fungi.filt.e<-filtered.qualy[which(filtered.qualy$e_value<=1e-50),]
dim(fungi.filt.e)

```

```{r}
# We excluded 279 OTUs because of e-value < 1e-50
nrow(filtered.qualy)-nrow(subset(filtered.qualy, e_value<=1e-50))

nrow(subset(filtered.qualy, e_value>1e-50&e_value<1e-20))

nrow(subset(filtered.qualy, e_value>1e-20))
```


```{r}
filtered.fungi.evalue <- subset(otu_table(phylo.fungi), rownames(otu_table(phylo.fungi)) %in% fungi.filt.e$OTU_ID...1)

phylo.fungi.evalue <- merge_phyloseq(filtered.fungi.evalue, tax_table(phylo.fungi), sample_data(phylo.fungi))

phylo.fungi.evalue
```

# Save phylo.data
```{r}
saveRDS(phylo.fungi.evalue, file = "2024-01-13 phylo.data.rarefaction.fungi.evalue.rds")
```

```{r}
OTUsum<-as.integer(taxa_sums(phylo.fungi.evalue))
str(OTUsum)
mod.oct <- prestonfit(OTUsum, tiesplit = F)
plot(mod.oct)

# sum(OTUsum) 
# length(OTUsum) 

```

Along this quality filtering I have excluded `r length(taxa_names(phylo.data))-length(taxa_names(phylo.fungi.evalue))` OTUs which represent the `r round((length(taxa_names(phylo.data))-length(taxa_names(phylo.fungi.evalue)))/length(taxa_names(phylo.data))*100,2)`% of the initial OTU table with Kingdom FUNGI.

## OTU length

Minimum OTU length is `r min(fungi.filt.e[,"query_len"])` and, maximum OTU length is `r max(fungi.filt.e[,"query_len"])` in this case.

Then, **I am not filtering short length OTUs**. 


```{r eval= FALSE}
write.table(otu_table(phylo.fungi.evalue), "2024-01-13 OTU Table after quality filtering.csv", append = FALSE, quote = TRUE, sep = ";", eol = "\n", na = "NA", dec = ".", row.names = FALSE,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")
```


# Alpha diversity measures - Nature: homogenezation paper *******************
```{r}
alpha_est <- data.frame(
  n_seq=colSums(otu_table(phylo.fungi.evalue)),
  Richness = estimate_richness(phylo.fungi.evalue , measures = "Observed")[,1],
  Chao1 = estimate_richness(phylo.fungi.evalue , measures = "Chao1")[,1],
  ACE = estimate_richness(phylo.fungi.evalue , measures = "ACE")[,1],
  Shannon = estimate_richness(phylo.fungi.evalue , measures = "Shannon")[,1],
  InvSimpson = estimate_richness(phylo.fungi.evalue , measures = "InvSimpson")[,1],
  Evenness = evenness(phylo.fungi.evalue,'pielou')
)

head(alpha_est)
summary(head(alpha_est))

corr <- round(cor(alpha_est), 1)
ggcorrplot(corr, hc.order = TRUE, type = "lower",
   lab = TRUE)


alpha_est%>%
  ggplot(aes(x=n_seq, y=Richness))+
  geom_point()+
  geom_smooth()

alpha_est%>%
  ggplot(aes(x=n_seq, y=Chao1))+
  geom_point()+
  geom_smooth()

```

```{r eval= FALSE}

phylo.fungi.evalue_sample<-sample_data(phylo.fungi.evalue)

phylo.fungi.evalue_sample<-merge(phylo.fungi.evalue_sample, alpha_est, by="row.names", all.x=TRUE)

head(phylo.fungi.evalue_sample)

dim(phylo.fungi.evalue_sample)

write.table(phylo.fungi.evalue_sample, "2024-01-13 Sample data after quality filtering.csv", append = FALSE, quote = TRUE, sep = ";", eol = "\n", na = "NA", dec = ".", row.names = FALSE,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")
```


```{r eval= FALSE}
write.table(tax_table(phylo.fungi.evalue), "2024-01-13 OTU Taxo after quality filtering.csv", append = FALSE, quote = TRUE, sep = ";", eol = "\n", na = "NA", dec = ".", row.names = FALSE,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")
```

